pkg/orchestrator/endpoint.go

```diff
@@ -60,6 +64,30 @@ func (e *BaseEndpoint) SubmitJob(ctx context.Context, request *SubmitJobRequest)
 		return nil, err
 	}
 
+	// Before we create an evaluation for the job, we want to check that none of the job's tasks
+	// need translating from a custom job type to a known job type (docker, wasm). If they do,
+	// then we will perform the translation and create the evaluation for the new job instead.
+	translatedJob, err := translation.Translate(ctx, e.taskTranslator, job)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we have translated the job (i.e. at least one task was translated) then we will switch
+	// to using the translated job after we have saved it in the jobstore.
+	//
+	// TODO: We want to return to the user the JobID of the original job, not the translated job.
+	// But this means the evaluation will not match as we only have an evaluation for the translated
+	// job. We need to think about how we can handle this or whether we have to move the translation
+	// to the compute node (where it will end up being run anyway).
```

Comment:
So what's the impact of the TODO today? It sounds like a `described` translated job will have no evaluations, executions or results?

Can we introduce job versioning? And the translated job can just be a subsequent version of the original, under the same ID? 

I would be in favour of porque no los dos, where compute node can advertise custom job types if it has them and the requester can just pass them along, but also the requester can perform the translation which is easier to manage in a big cluster.

Reply:
Yes, the original (exec python) job will have no evaluations, or execution or results, it exists solely to act as a record for the newly translated (docker run python) job.  The translated job contains a metadata reference back to the original job, but this is only one way - we can only save a job once, after which we only update it's state.  

We already have version and revision field in the job model. Revisions are used to check sequence/ordering when interacting with the jobstore, but Versions are mostly unused at present.  We _could_ start making use of them but I suspect it won't be a short piece of work as we'll need to introduce new jobstore functionality for updating jobs (we can't currently) and changes to the interface for retrieving different versions. 

We _could_ ensure the metadata reference is bidirectional (by late saving the origin job) and using that in calls to get/describe to retrieve the origin job specification?